Add support for string nodes to `StaticLayoutProvider`

mattpap opened the pull request:
This PR allows to use string nodes with `StaticLayoutProvider`. Previously only integer keys were allowed. In future this could potentially be generalized to any types, but is out of scope for this particular work. As a side effect of this change, I had to figure out a solution to issue #13378, i.e. resolve deserialization issues of dictionaries. This PR partially supersedes PR #13437. I will refocus that PR from attempting to fix the issue to its actual main point, which is migration from plain objects to `Map` type. This PR also removes some leftovers after removal of hammerjs, which I noticed when dealing with some type issues.

fixes #12651 
fixes #13378

mattpap committed:
Add support for string nodes to StaticLayoutProvider

mattpap committed:
Remove plain object serialization/deserialization hacks

mattpap committed:
Remove leftovers after removal of hammerjs

mattpap committed:
Remove custom plain object to Map conversions

mattpap committed:
Make HelpButton.icon override lazily evaluated

mattpap committed:
Allow PlainObject<T> properties to accept Map<string, T>

mattpap committed:
Handle changes to ColumnarDataSource.data's type

mattpap committed:
Change the type of Selection.multiline_indices to Map

mattpap committed:
Update graph hit test policies and layout providers

mattpap committed:
Change the type of CDSView.indices_map to Map

mattpap committed:
Update DataTable and DataCube

mattpap committed:
Remove unneeded Kind.coerce()

mattpap committed:
Update bokehjs' tests

mattpap committed:
Update bokehjs' examples

mattpap committed:
Improve type of WebDataSource.adapter

mattpap committed:
Return Map from ColumnarDataSource.inferred_defaults

mattpap committed:
Preserve data type in ColumnarDataSource.clear()

mattpap committed:
Finalize updating bokehjs' unit tests

mattpap commented:
The original half-measure approach in this PR didn't fully work, so I had to do things properly. This means allowing `Map<string, T> | {[key: string}: T}` where previously only `{[key: string}: T}` was allowed. This means no more hacks or implicit type coercions of any kind. Internally all properties that accept this border type use a proxying mechanism, so that all relevant code can assume `Map` interface and use `get()` or `set()` instead of index signatures and dreadful JS `in` operator. This in effect means that missing keys are finally handled properly.

This approach is different from what I attempted in PR #13437. There I converted plain objects to `Map` and tried to use asymmetric setters and getters to tie that on the type-level. This didn't work well, due to mutability of objects and that setters and getters with different types don't integrate well with the rest of TypeScript.

mattpap committed:
Adjust handling of TapTool.modifiers

bryevdv commented on src/bokeh/models/graphs.py:
```diff
@@ -82,11 +85,11 @@ class StaticLayoutProvider(LayoutProvider):
-    The coordinates of the graph nodes in cartesian space. The dictionary
-    keys correspond to a node index and the values are a two element sequence
-    containing the x and y coordinates of the node.
+    graph_layout = Dict(Either(Int, String), OfLength(Seq(Float), 2), default={}, help="""
```
The order of `OfLength(Seq(Float), 2)` makes for extremely awkward grammar. At a minimum, switching the order to `OfLength(2, Seq(Float))` would read more clearly, unless something like`Seq(Float).OfLength(2)` is attainable.

mattpap replied:
> `OfLength(Seq(Float), 2)`

This is not far off from `len(seq) == 2`, which is what I was targeting. Maybe I should have used `Len(Seq(Float), 2)` instead? However, adding a method to container property type seems a very good idea.

bryevdv replied:
`Len` would be preferable to `OfLength` (to me at least) for a quick and simple rename option

bryevdv commented on examples/server/app/population.py:
```diff
@@ -24,10 +24,8 @@
-            val[i] = -xs[i]
-        else
-            val[i] = xs[i]
+        const sign = source.get("Sex")[i] == "Male" ? -1 : 1
```
The old code still functions and this is just by way of improvement (here and other examples)?

mattpap replied:
It did not. However, taking advantage of the fact that `ColumnDataSource.data` is wrapped in `PropertyValueColumnData`, I was able to add a special case to preserve compatibility. Otherwise in Python we don't know what is a true dict and what is a pseudo-struct encoded as dict. In bokehjs types and object identities are preserved, so that equality and mutability still work as expected.

mattpap replied:
I reverted changes to bokeh's examples.

bryevdv commented on bokehjs/test/unit/models/tools/edit/poly_edit_tool.ts:
```diff
@@ -245,11 +253,15 @@ describe("PolyEditTool", (): void => {
+        xs: [[0, 0.5, 1], [0, 0.4646017699115044, 1]],
+        ys: [[0, -0.5, -1], [0, -0.4661016949152542, -1]],
+        z: [null, null],
+      })
```
Why is this form preferable? Presumably split out as they were before affords finer-grained test failure messages

mattpap replied:
I prefer to have full context of a data structure, to be able to see broader failure pattern, which testing one line at a time doesn't allow. It also forces the writer to deal with all fields and there were some missing in other tests.  It would be optimal to pair this with adding support for data structure diff to bokehjs' testing framework, which I have intended to add for a long while.

mattpap committed:
Add special handling for PropertyValueColumnData

mattpap committed:
Rename OfLength(seq, n) -> Len(seq, n)

mattpap committed:
Update defaults' and cross' baselines

mattpap committed:
Update advanced/extensions/wrapping.py

This is needed for type checking purpose. The data is
a plain object as before.

mattpap committed:
Ignore `plain` attribute when comparing `{type: "map"}`
